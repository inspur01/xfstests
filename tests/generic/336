#! /bin/bash
# FSQA Test No. 336
#
# Test that if we have a file F1 with two links, one in a directory A and the
# other in directory B, if we remove the link in directory B, move some other
# file F2 from directory B into directory C, fsync inode F1, power fail and
# remount the filesystem, file F2 exists and is located only in directory C.
#
#-----------------------------------------------------------------------
#
# Copyright (C) 2016 SUSE Linux Products GmbH. All Rights Reserved.
# Author: Filipe Manana <fdmanana@suse.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it would be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write the Free Software Foundation,
# Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#-----------------------------------------------------------------------
#

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"
tmp=/tmp/$$
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	_cleanup_flakey
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/dmflakey

# real QA test starts here
_supported_fs generic
_supported_os Linux
_require_scratch
_require_dm_target flakey
_require_metadata_journaling $SCRATCH_DEV

rm -f $seqres.full

_scratch_mkfs >>$seqres.full 2>&1
_init_flakey
_mount_flakey

# Create our test directories and the file we will later check if it has
# disappeared (file bar).
mkdir $SCRATCH_MNT/a
mkdir $SCRATCH_MNT/b
mkdir $SCRATCH_MNT/c
touch $SCRATCH_MNT/a/foo
ln $SCRATCH_MNT/a/foo $SCRATCH_MNT/b/foo_link
touch $SCRATCH_MNT/b/bar

# Make sure everything is durably persisted.
sync

# Now delete one of the hard links of file foo and move file bar into c/
unlink $SCRATCH_MNT/b/foo_link
mv $SCRATCH_MNT/b/bar $SCRATCH_MNT/c/

# Now fsync file foo.
$XFS_IO_PROG -c "fsync" $SCRATCH_MNT/a/foo

echo "Filesystem content before power failure:"
ls -R $SCRATCH_MNT/a $SCRATCH_MNT/b $SCRATCH_MNT/c | _filter_scratch

# Simulate a power failure / crash and remount the filesystem, so that the
# journal/log is replayed.
_flakey_drop_and_remount

# We expect that after the journal/log was replayed, we no longer have the link
# foo_link and file bar was moved from directory b/ to directory c/.
echo "Filesystem content after power failure:"
# Must match what we had before the power failure.
ls -R $SCRATCH_MNT/a $SCRATCH_MNT/b $SCRATCH_MNT/c | _filter_scratch

_unmount_flakey

status=0
exit
