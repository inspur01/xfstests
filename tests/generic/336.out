QA output created by 336
Filesystem content before power failure:
SCRATCH_MNT/a:
foo

SCRATCH_MNT/b:

SCRATCH_MNT/c:
bar
Filesystem content after power failure:
SCRATCH_MNT/a:
foo

SCRATCH_MNT/b:

SCRATCH_MNT/c:
bar
